libr/core/windows_heap.c

```diff
@@ -191,26 +226,62 @@ static void free_extra_info(PDEBUG_HEAP_INFORMATION heap) {
 	}
 }
 
-static bool DecodeHeapEntry(PHEAP heap, PHEAP_ENTRY entry) {
-#if defined(_M_X64)
-	(WPARAM)entry += sizeof (PVOID);
-#endif
+static WPARAM GetNtDllOffset(RDebug *dbg) {
+	RList *map_list = r_w32_dbg_maps (dbg);
+	RListIter *iter;
+	RDebugMap *map;
+	WPARAM ntdllOffset = 0;
+	// Get ntdll .data location
+	r_list_foreach (map_list, iter, map) {
+		if (strstr (map->name, "ntdll.dll | .data")) {
+			ntdllOffset = map->addr;
+			break;
+		}
+	}
+	r_list_free (map_list);
+	return ntdllOffset;
+}
+
+static WPARAM GetLFHKey(RDebug *dbg, HANDLE h_proc, bool segment) {
+	WPARAM lfhKey = 0;
+	WPARAM lfhKeyLocation;
+	WPARAM ntdllOffset = GetNtDllOffset (dbg);
+
+	if (ntdllOffset) {
+		if (segment) {
+			WPARAM RtlpHpHeapGlobalsOffset = ntdllOffset + 0x44A0; // ntdll!RtlpHpHeapGlobals
+			lfhKeyLocation = RtlpHpHeapGlobalsOffset + sizeof (WPARAM);
+		} else {
+			lfhKeyLocation = ntdllOffset + 0x7508; // ntdll!RtlpLFHKey
+		}
+		if (!ReadProcessMemory (h_proc, (PVOID)lfhKeyLocation, &lfhKey, sizeof (WPARAM), NULL)) {
+			r_sys_perror ("ReadProcessMemory");
+			eprintf ("LFH key not found.\n");
+		}
+	}
+	return lfhKey;
+}
+
+static bool DecodeHeapEntry(PHEAP heap, PHEAP_ENTRY entry, int bitness) {
+	if (bitness == R_SYS_BITS_64) {
+		(WPARAM) entry += sizeof (PVOID);
+	}
 	if (heap->EncodeFlagMask && (*(UINT32 *)entry & heap->EncodeFlagMask)) {
-#if defined(_M_X64)
-		(WPARAM)heap += sizeof (PVOID);
-#endif
+		if (bitness == R_SYS_BITS_64) {
```

Comment:
While you are on it - imagine the scenario where radare2 is built for the Linux i386 machine, or ARMv7, while connecting to the Windows 10 64bit through `windbg://` remote debugging protocol. Thus the heap decoding should not depend on the host (radare2) parameters, but rather the parameters set with `asm.arch`, `asm.bits`, etc. See the example of our TCC-based C types definitions parsing code, where the size of `long` depends on the architecture and OS: https://github.com/radare/radare2/blob/master/shlr/tcc/tccgen.c#L515

Reply:
This code only runs in target machine. So the remote debugging situation wont be possible to happen

Reply:
@radare the idea is to be able to parse heap remotely as well, or even from MDMP files. So why not to lay the ground right now?